pkg/cinder/ceph_backend.go

```diff
@@ -0,0 +1,118 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cinder
+
+import (
+	"bytes"
+	cinderv1beta1 "github.com/openstack-k8s-operators/cinder-operator/api/v1beta1"
+	"net"
+	"sort"
+	"strings"
+	"text/template"
+)
+
+type CephDefaults string
+
+type Spec struct {
+	CephDefaults *CephDefaults
+}
+
+type T struct {
+	Pool string
+}
+
+// Cinder defaults
+const (
+	CephDefaultCinderUser CephDefaults = "openstack"
+	CephDefaultCinderPool CephDefaults = "volumes"
+)
+
+// GetCephBackend is a function that validates the ceph client parameters.
+func GetCephBackend(instance *cinderv1beta1.Cinder) bool {
+	// Validate parameters to enable CephBackend
+	if instance.Spec.CephBackend.CephClusterFSID != "" &&
+		validateMons(instance.Spec.CephBackend.CephClusterMonHosts) &&
+		instance.Spec.CephBackend.CephClientKey != "" {
+		return true
+	}
+	return false
+}
+
+// This is a utily function that validates the comma separated Mon list defined
+// for the external ceph cluster; it also checks the provided IP addresses are not
+// malformed
+func validateMons(ipList string) bool {
+	for _, ip := range strings.Split(ipList, ",") {
+		if net.ParseIP(strings.Trim(ip, " ")) == nil {
+			return false
+		}
+	}
+	return true
+}
+
+// GetCephCinderPool is a function that validate the pool passed as input, and return
+// volumes no pool is given
+func GetCephCinderPool(instance *cinderv1beta1.Cinder) string {
+
+	if pool, found := instance.Spec.CephBackend.CephPools["cinder"]; found {
+		return pool.CephPoolName
+	}
+	return string(CephDefaultCinderPool)
+}
+
+// GetCephRbdUser is a function that validate the user passed as input, and return
+// openstack if no user is given
+func GetCephRbdUser(instance *cinderv1beta1.Cinder) string {
+
+	if instance.Spec.CephBackend.CephUser == "" {
+		return string(CephDefaultCinderUser)
+	}
+	return instance.Spec.CephBackend.CephUser
+}
+
+// GetCephOsdCaps is a function that returns the Caps for each defined pool
+func GetCephOsdCaps(instance *cinderv1beta1.Cinder) string {
+
+	var osdCaps string // the resulting string containing caps
+	t, _ := template.New("osdCaps").Parse("profile rbd={{ .Pool }},")
+
+	/**
+	A map of strings (pool service/name in this case) is, by definition, an
+	unordered structure, and let the function return a different pattern
+	each time. This causes the ConfigMap hash to change, and the pod being
+	redeployed because the operator detects the different hash. Sorting the
+	resulting array of pools makes everything predictable
+	**/
+	var plist []string
+	for _, pool := range instance.Spec.CephBackend.CephPools {
+		plist = append(plist, pool.CephPoolName)
+	}
+	// sort the pool names
+	sort.Strings(plist)
+
+	// Build the resulting caps from the _ordered_ array applying the template
+	for _, pool := range plist {
+
+		osdCaps += func(p string) string {
+			var d bytes.Buffer
+			t.Execute(&d, T{p})
+			return (d.String())
+		}(pool)
+	}
```

Comment:
Just curious as to why a template was needed here.  `fmt.Sprintf()` seems like it might work and would be simpler, but perhaps I am missing something.

Reply:
mmm I guess it's just a different approach to achieve the same result :D

Reply:
Ok, simplifying this part adding fmt.Sprintf

Reply:
Done.